### PR TITLE
PR for #40 - Add C++ language

### PR DESCRIPTION
**Pull request by chess-equality**
#40

### CODE REVIEW

File: src/main/groovy/com/codebrig/arthur/observe/structure/filter/operator/misc/TernaryOperatorFilter.groovy [hunk 1]
```diff
@@ -25,14 +26,15 @@ class TernaryOperatorFilter extends StructureFilter<TernaryOperatorFilter, Void>
...
+                        new RoleFilter("CONDITION"), new RoleFilter("EXPRESSION"),
+                        new InternalRoleFilter().reject("Prop_InitOperand2")
                 )
         )
     }
 
     @Override
     boolean evaluate(SourceNode node) {
-        if (node.internalType == "ConditionalExpression") {
+        if (Objects.requireNonNull(node).internalType == "ConditionalExpression") {
```
Comment (BFergerson) [inline src/main/groovy/com/codebrig/arthur/observe/structure/filter/operator/misc/TernaryOperatorFilter.groovy]: You can't require non-null here. `node` is allowed to be null.

File: src/main/groovy/com/codebrig/arthur/observe/structure/filter/operator/relational/compare/IsEqualOperatorFilter.groovy [hunk 2]
```diff
@@ -27,6 +28,10 @@ class IsEqualOperatorFilter extends StructureFilter<IsEqualOperatorFilter, Void>
 
     @Override
     boolean evaluate(SourceNode node) {
-        return filter.evaluate(node)
+        if (Objects.requireNonNull(node).internalType == "CPPASTBinaryExpression" && node.token == "==") {
```
Comment (BFergerson) [inline src/main/groovy/com/codebrig/arthur/observe/structure/filter/operator/relational/compare/IsEqualOperatorFilter.groovy]: Same thing. Filters must check that it's not-null if that's part of its logic, but they can't require it not to be null.

File: src/main/groovy/com/codebrig/arthur/observe/structure/filter/operator/misc/TernaryOperatorFilter.groovy (hunk 1 above)
Comment (chess-equality) [reply src/main/groovy/com/codebrig/arthur/observe/structure/filter/operator/misc/TernaryOperatorFilter.groovy]: @BFergerson Please see my latest commit https://github.com/CodeBrig/Arthur/pull/86/commits/058027cc22beaf457f932ab399abe9aced5e5a52.

**Comment by BFergerson** [comment]
@chess-equality, please add the missing `SameProgram.cpp`

**Comment by chess-equality** [comment]
@BFergerson Please see my commit https://github.com/chess-equality/Arthur/commit/d0c2e4d854d5ac8ac9fdd4114e998fa1a8e1f96d

**Comment by BFergerson** [comment]
Opened some PRs to deal with missing roles:
 - https://github.com/bblfsh/cpp-driver/pull/57
 - https://github.com/bblfsh/cpp-driver/pull/58
 - https://github.com/bblfsh/cpp-driver/pull/59

Will keep this PR open till those are released.

**Comment by BFergerson** [comment]
Added todos to track PRs. It doesn't seem babelfish updates its drivers all too often. Will use the current way until they release new versions.

**Comment by chess-equality** [comment]
> Added todos to track PRs. It doesn't seem babelfish updates its drivers all too often. Will use the current way until they release new versions.

@BFergerson Ok, thanks for heads-up.

**Pull request merged**